src/libtools/property.cc

```diff
@@ -179,88 +177,87 @@ void PrintNodeTXT(std::ostream &out, Property &p, const int start_level, int lev
-            // print node value if it is not empty
-            has_value = ( (p.value()).find_first_not_of("\t\n ") != std::string::npos );
-            if( has_value ) { 
-                out <<  cAttributeValue <<  p.value() << cReset;
-                _endl = false; 
-            }
-            
-            // check if we need the end of the line or not
-            if( !has_value &&  p.HasChilds() ) out << endl;
-            if( !has_value &&  !p.HasChilds() ) _endl = false;
+      }
+
+      string cKey = color->Magenta();
+      string cAttribute = color->Blue();
+      string cAttributeValue = color->Green();
+      string cReset = color->Reset();
+
+      // print starting only from the start_level (the first node (level 0) can be <> </>)
+      if (level >= start_level) {
+        // print the node name
+        out << indent << offset << "<" << cKey << p.name() << cReset;
+        // print the node attributes 
+        for (ia = p.firstAttribute(); ia != p.lastAttribute(); ++ia) {
+          out << " "
+                  << cAttribute << ia->first << cReset
+                  << "=\""
+                  << cAttributeValue << ia->second << cReset << "\"";
         }
-    
-        // continue iteratively through the rest of the nodes
-        for(iter = p.begin(); iter!=p.end(); ++iter) {
-            level++; 
-            if (  level > start_level ) offset += "\t"; 
-            PrintNodeXML(out, (*iter), piom, level, offset);
-            if (  level > start_level ) offset.resize(offset.size()-1); 
-            level--;           
+        // print node value if it is not empty
+        has_value =  ((p.value()).find_first_not_of("\t\n ") != std::string::npos);
+        if (has_value || p.HasChilds()) {
+          out << ">";
+        } else {
+          out << "/>"<<std::endl;
         }
-        
-        if ( level >= start_level ) {
-            if ( _endl ) {
-                         out << indent << offset << "</" << cKey << p.name() << cReset << ">" << endl;
-            } else {
-                if (has_value) { // avoid indent after the value
-                        out << "</" << cKey << p.name() << cReset << ">" << endl;
-                } else {
-                        out << indent << offset << "</" << cKey << p.name() << cReset << ">" << endl;
-                }
-            }
-        } 
-}
+        if (has_value) {
+          out << cAttributeValue << p.value() << cReset;
+          linebreak = false;
+        }
+
+        // check if we need the end of the line or not
+        if (!has_value && p.HasChilds()) out << std::endl;
+        if (!has_value && !p.HasChilds()) linebreak = false;
+      }
+
+      // continue iteratively through the rest of the nodes
+      for (Property& prop : p) {
+        level++;
+        if (level > start_level) offset += "\t";
+        PrintNodeXML(out, prop, piom, level, offset);
+        if (level > start_level) offset.resize(offset.size() - 1);
+        level--;
+      }
```

Comment:
This doesn't really matter but just for your info: I believe preincremeinting is supposed to be faster than post though some people argue it doesn't really matter. 
https://softwareengineering.stackexchange.com/questions/358725/pre-increment-vs-post-increment

```
--level;
++level;
```

Reply:
for simple data types it makes no difference